### ISSUE TITLE
"Get Binary File" keyword consumes high memory

### ISSUE THREAD

**Issue by chidambaranathan-r** [robotframework/robotframework#4658]
"`Get Binary File`" keyword consumes high memory in proportional to the size of the file passed as input to it.

File size of approximately ~270MB is used as an input. Below is the memory consumption (~12GB) shown as reference:

![memory_size](https://user-images.githubusercontent.com/1535689/218940257-495a780b-1deb-492c-865a-ea7959c9d8a8.jpg)

This has an adverse effect when the same keyword is called in multiple test cases that are running in parallel using `pabot`, which leads to "oom-kill" from kernel.

Similar issue is reported here (but it was closed due to inactive thread): https://github.com/robotframework/robotframework/issues/3827

This is reproducible in local laptop itself by giving an input file of larger size "Get Binary File" keyword. To invoke "oom-kill", same keyword (with larger file) can be invoked in parallel.

Can there be new keyword introduced "to upload of large files without reading them into memory" like ? or Is there any keyword that does this job already?

I have detailed this in (where we suspected a different keyword but it turned out to be `Get Binary File`): https://groups.google.com/g/robotframework-users/c/gUPNXWq-Nvc

**Comment by pekkaklarck** [comment]
`Get Binary File` is designed to read the whole file into memory. That obviously won't work well if the file is bigger than the available memory. I don't think we can do anything to that.

If you are uploading the file, doing that without reading the whole fine into memory is obviously a good idea. Robot's standard library doesn't have any keywords for uploading or downloading, but I'm sure there are external libraries that support it and as the last resort you can always create your own custom keyword.

Adding a generic uploading/downloading library to Robot could be considered, but it would need someone to lead the effort to implement it. It would be best to start with an external library, but if it would use only Python's standard library, it could be integrated with Robot later. Is this something you are interested in?

**Comment by chidambaranathan-r** [comment]
Thanks for the response @pekkaklarck

We found a keyword called [Get File For Streaming Upload](https://marketsquare.github.io/robotframework-requests/doc/RequestsLibrary.html#Get%20File%20For%20Streaming%20Upload) from external robot framework library called [RequestsLibrary](https://github.com/MarketSquare/robotframework-requests) which works well for our case, and it consumes almost no memory.

**Comment by pekkaklarck** [comment]
Apparently that keyword just opens a file and returns the open file object (basically a handle) where other keywords can read the data. Most likely then the keyword actually uploading the data will read the content in pieces instead of reading the file fully into memory.

We could consider adding keywords for opening binary and text files to the OperatingSystem. They'd basically be wrappers to the built-in [open()](https://docs.python.org/3/library/functions.html#open). Something like this:
```python
def open_file(path, mode='r', encoding='UTF-8', errors='strict'):
    return open(path, mode, encoding=encoding, errors=errors)

def open_binary_file(path, mode='rb'):
    return open(path, mode)
```
These keywords wouldn't be useful on their own, so we needed to have also keywords that use the open files. Existing reading/writing keywords could probably be enhanced to support open file objects, but a new keyword would be needed for closing opened files. There possibly could be some auto-closing functionality as well to close opened files automatically when a test or a suite ends.

Such lower level keywords wouldn't be that useful on their own compared to simply reading and writing files. They could, however, be used with other keywords for processing files more efficiently. If you think keywords for opening files would be useful, please submit a new issue. There isn't anything we can do for `Get Binary File` on its own.
